share/steel/examples/pulse/lib/PulseHashTable.fst

```diff
@@ -0,0 +1,284 @@
+    ))
+  {
+    let voff = !off;
+    assume_ (pure (US.fits (US.v cidx + US.v voff)));
+    if (voff = ht.sz) {
+      cont := false;
+      assert (R.pts_to ret full_perm None);
+    } else {
+      let idx = modulo_us cidx voff ht.sz ();
+      let c = op_Array_Index ht.contents idx; 
+      match c {
+        Used k' v' -> {
+          if (k' = k) {
+            cont := false;
+            write ret (Some (v',idx)) #None;
+            assert (pure (pht.repr @@ US.v idx == Used k' v'));
+            assert (pure (lookup_repr_index pht.repr k == Some (v', US.v idx)));
+          } else {
+            off := voff + 1sz;
+            assert (pure (walk_get_idx pht.repr (US.v cidx) k (US.v voff) 
+              == walk_get_idx pht.repr (US.v cidx) k (US.v (voff+1sz))));
+          }}
+        Clean -> {
+          cont := false;
+          assert (pure (walk_get_idx pht.repr (US.v cidx) k (US.v voff) == None));
+          assert (R.pts_to ret full_perm (PHT.lookup_index_us pht k));
+         }
+        Zombie -> {
+          off := voff + 1sz;
+          assert (pure (walk_get_idx pht.repr (US.v cidx) k (US.v voff) 
+            == walk_get_idx pht.repr (US.v cidx) k (US.v (voff+1sz))));
+         }
+      };
+    };
+  };
+  fold (models s ht pht);
+  assert (R.pts_to ret full_perm (PHT.lookup_index_us pht k));
+  let o = !ret;
+  o
+}
+```
+
+```pulse
+fn _lookup (#s:pht_sig_us)
+          (#pht:erased (pht_t (s_to_ps s)))
+          (ht:ht_t s) (k:s.keyt)
+  requires models s ht pht
+  returns  o:(o:option s.valt{o == PHT.lookup pht k})
+  ensures  models s ht pht
+{
+  let o = pulse_lookup_index #s #pht ht k;
+  match o {
+    Some p -> { Some (fst p) }
+    None -> { None #s.valt }
+  }
+}
+```
+let lookup #s #pht ht k = _lookup #s #pht ht k
+
+```pulse
+fn _insert (#s:pht_sig_us)
+          (#pht:(p:erased (pht_t (s_to_ps s)){PHT.not_full p.repr}))
+          (ht:ht_t s) (k:s.keyt) (v:s.valt)
+  requires models s ht pht
+  ensures  models s ht (PHT.insert pht k v)
+{
+  let cidx = canonical_index_us k ht.sz;
+  let mut off = 0sz;
+  let mut cont = true;
+
+  unfold (models s ht pht);
+
+  while (let voff = !off; let vcont = !cont; (voff <= ht.sz && vcont = true)) 
+  invariant b. exists (voff:US.t) (vcont:bool). (
+    R.pts_to off full_perm voff **
+    R.pts_to cont full_perm vcont **
+    A.pts_to ht.contents full_perm (if vcont then pht.repr else (PHT.insert pht k v).repr) **
+    pure (
+      US.v ht.sz == pht_sz pht /\
+      voff <= ht.sz /\
+      strong_all_used_not_by pht.repr (US.v cidx) (US.v voff) k /\
+      walk pht.repr (US.v cidx) k (US.v voff) == lookup_repr pht.repr k /\
+      insert_repr_walk #(s_to_ps s) #(pht_sz pht) #pht.spec pht.repr k v (US.v voff) (US.v cidx) () () 
+        == insert_repr #(s_to_ps s) #(pht_sz pht) #pht.spec pht.repr k v /\
+      b == (voff <= ht.sz && vcont = true)
+    ))
+  {
+    let voff = !off;
+    assume_ (pure (US.fits (US.v cidx + US.v voff)));
```

Comment:
I see this assume in a few places. Can you describe what the issue is and what we need to do to solve it?

Reply:
depending on how large the hash table is, we technically don't know that sum of the canonical index and the offset (both of which are less than or equal to the hash table size) fits within a machine integer. e.g. consider the case where the hash table size is the largest machine integer. so we need this assume (or we could alternatively assume that the hash table size is less than half the largest machine integer, and I think I could get that to check)

Reply:
One common way to do this is to have a dynamic check for overflow. E.g., one could write a function to safely add cidx and voff, e.g., by first comparing if SIZE_T_MAX - cidx > voff and then doing the addition. If the check fails, either the whole thing can fail, or better, insert returns a boolean to signal if the insert actually happened, or if overflow was detected and the insert was skipped.

Reply:
on it!